### ISSUE TITLE
Tray module not visible after some time. No such object path '/StatusNotifierWatcher'

### ISSUE THREAD

**Issue by HarshNarayanJha** [Alexays/Waybar#3468]
After a few hours of logging into hyprland, the Tray Module does now show up after killing and launching waybar.

This does not happen before that point, I can kill waybar and start it again (via a keybind) and the tray and all its icons are still there. But after some time has passed, reloading waybar no longer shows the Tray Module, no matter how many times I reload. 
But if I log out and login again, it immediately fixes the issue and the tray stays visible until the issue happens again.
Note that it never disappears automatically, only if I restart waybar.

Running waybar in terminal reveals the issue
```bash
...
[2024-07-23 19:38:36.383] [debug] Creating workspace 1
[2024-07-23 19:38:36.383] [debug] Creating workspace 2
[2024-07-23 19:38:36.496] [error] Host: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: No such object path '/StatusNotifierWatcher'
[2024-07-23 19:38:36.613] [info] Bar configured (width: 1920, height: 40) for output: 
...
```

EDIT: Following issue #3040, I have downgraded `libappindicator-gtk3` to `v12.10.0.r296-1` with no success

<details>
<summary>Full Log</summary>

```bash
[2024-07-23 19:38:58.575] [debug] Try expanding: $XDG_CONFIG_HOME/waybar/config
[2024-07-23 19:38:58.575] [debug] Try expanding: $XDG_CONFIG_HOME/waybar/config.jsonc
[2024-07-23 19:38:58.576] [debug] Try expanding: $HOME/.config/waybar/config
[2024-07-23 19:38:58.576] [debug] Found config file: $HOME/.config/waybar/config
[2024-07-23 19:38:58.576] [info] Using configuration file /home/harshnj/.config/waybar/config
[2024-07-23 19:38:58.579] [info] Unable to receive desktop appearance: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface “org.freedesktop.portal.Settings” on object at path /org/freedesktop/portal/desktop
[2024-07-23 19:38:58.579] [debug] Try expanding: $XDG_CONFIG_HOME/waybar/style.css
[2024-07-23 19:38:58.579] [debug] Try expanding: $HOME/.config/waybar/style.css
[2024-07-23 19:38:58.579] [debug] Found config file: $HOME/.config/waybar/style.css
[2024-07-23 19:38:58.579] [info] Using CSS file /home/harshnj/.config/waybar/style.css
[2024-07-23 19:38:58.580] [debug] Parsing imports for file: /home/harshnj/.config/waybar/style.css
[2024-07-23 19:38:58.580] [debug] Try expanding: $XDG_CONFIG_HOME/waybar/themes/theme.css
[2024-07-23 19:38:58.580] [debug] Try expanding: $HOME/.config/waybar/themes/theme.css
[2024-07-23 19:38:58.580] [debug] Found config file: $HOME/.config/waybar/themes/theme.css
[2024-07-23 19:38:58.580] [debug] Adding file to watch list: Catppuccin-Mocha.css
[2024-07-23 19:38:58.580] [debug] Adding file to watch list: /home/harshnj/.config/waybar/style.css
[2024-07-23 19:38:58.587] [debug] Output detection done: eDP-1 (AU Optronics 0x229E)
[2024-07-23 19:38:58.592] [info] Hyprland IPC starting
[2024-07-23 19:38:58.594] [debug] Initializing workspaces
[2024-07-23 19:38:58.595] [debug] Initializing persistent workspaces
[2024-07-23 19:38:58.595] [info] Loading persistent workspaces from Waybar config
[2024-07-23 19:38:58.595] [info] Loading persistent workspaces from Hyprland workspace rules
[2024-07-23 19:38:58.596] [info] Registering for Hyprland's 'windowtitle' events because a user-defined window rewrite rule uses the 'title' field.
[2024-07-23 19:38:58.616] [debug] network: new default route via 192.168.250.102 on if4 metric 600
[2024-07-23 19:38:58.616] [debug] network: wlan0, new addr 192.168.250.94/24
[2024-07-23 19:38:58.616] [debug] network: wlan0, new addr 192.168.250.94/24
[2024-07-23 19:38:58.641] [debug] Received SIGCHLD in signalThread
[2024-07-23 19:38:58.641] [debug] Cmd exited with code 0
[2024-07-23 19:38:58.987] [debug] GTK widget tree:
window#waybar.background.top.eDP-1..mode-default.solo.org.kde.konsole:dir(ltr)
  decoration:dir(ltr)
  box.horizontal:dir(ltr)
    box.horizontal.modules-left:dir(ltr)
      widget:dir(ltr)
        label#custom-l_end.module:dir(ltr)
      widget:dir(ltr)
        box#workspaces.horizontal.module:dir(ltr)
      widget:dir(ltr)
        box#window.horizontal.module:dir(ltr)
          image:dir(ltr)
          label:dir(ltr)
      widget:dir(ltr)
        label#custom-r_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-padd.module:dir(ltr)
    box.horizontal.modules-center:dir(ltr)
      widget:dir(ltr)
        label#custom-padd.module:dir(ltr)
      widget:dir(ltr)
        label#custom-l_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-weather.module:dir(ltr)
      widget:dir(ltr)
        label#custom-r_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-padd.module:dir(ltr)
    box.horizontal.modules-right:dir(ltr)
      widget:dir(ltr)
        label#custom-padd.module:dir(ltr)
      widget:dir(ltr)
        label#custom-l_end.module:dir(ltr)
      widget:dir(ltr)
        box#hardware.horizontal:dir(ltr)
          revealer.drawer:dir(ltr)
            box.horizontal:dir(ltr)
              widget.drawer-child:dir(ltr)
                label#cpu.module:dir(ltr)
              widget.drawer-child:dir(ltr)
                label#temperature.module:dir(ltr)
          widget:dir(ltr)
            label#memory.module:dir(ltr)
      widget:dir(ltr)
        label#custom-r_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-l_end.module:dir(ltr)
      widget:dir(ltr)
        label#mpris.module:dir(ltr)
      widget:dir(ltr)
        label#backlight.module:dir(ltr)
      widget:dir(ltr)
        label#network.module:dir(ltr)
      widget:dir(ltr)
        label#pulseaudio.module:dir(ltr)
      widget:dir(ltr)
        label#pulseaudio.module.microphone:dir(ltr)
      widget:dir(ltr)
        label#custom-r_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-l_end.module:dir(ltr)
      widget:dir(ltr)
        box#privacy.horizontal:dir(ltr)
          revealer:dir(ltr)
            box#privacy-item.horizontal.screenshare:dir(ltr)
              image:dir(ltr)
          revealer:dir(ltr)
            box#privacy-item.horizontal.audio-out:dir(ltr)
              image:dir(ltr)
          revealer:dir(ltr)
            box#privacy-item.horizontal.audio-in:dir(ltr)
              image:dir(ltr)
      widget:dir(ltr)
        box#gamemode.horizontal:dir(ltr)
          image:dir(ltr)
          label:dir(ltr)
      widget:dir(ltr)
        box#tray.horizontal.module:dir(ltr)
      widget:dir(ltr)
        label#battery.module:dir(ltr)
      widget:dir(ltr)
        label#custom-r_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-l_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-notification.module:dir(ltr)
      widget:dir(ltr)
        label#idle_inhibitor.module:dir(ltr)
      widget:dir(ltr)
        label#clock.module:dir(ltr)
      widget:dir(ltr)
        label#custom-r_end.module:dir(ltr)
      widget:dir(ltr)
        label#custom-l_end.module:dir(ltr)
      widget:dir(ltr)
        box#user.horizontal.module:dir(ltr)
          image:dir(ltr)
          label:dir(ltr)
      widget:dir(ltr)
        label#custom-power.module:dir(ltr)
      widget:dir(ltr)
        label#custom-r_end.module:dir(ltr)

[2024-07-23 19:38:58.989] [debug] Creating workspace 1
[2024-07-23 19:38:58.990] [debug] Creating workspace 2
[2024-07-23 19:38:59.103] [error] Host: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: No such object path '/StatusNotifierWatcher'
[2024-07-23 19:38:59.196] [info] Bar configured (width: 1920, height: 40) for output: eDP-1
[2024-07-23 19:38:59.221] [debug] hyprland IPC received openlayer>>waybar
[2024-07-23 19:38:59.869] [debug] Received SIGCHLD in signalThread
[2024-07-23 19:38:59.869] [debug] Cmd exited with code 0
```

</details>

**Comment by HarshNarayanJha** [comment]
Wait why these tags bot? Beacuse of the log?

**Comment by ShikharArvind** [comment]
- Similar issue on Sway + Waybar. Resuming after extended suspend makes the tray part of bar disappear. Killing and restarting it does not help. 
- Restarting Sway itself fixes the issue, and the tray seems to be working fine. 
- Starting from terminal throws the error `Host: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: No such object path '/StatusNotifierWatcher'`
- Could be due to `libappindicator-gtk3` as described in issue #3040 .  Version is 12.10.0.r298-4 for me.  Maybe downgrading to 12.10.0.r296 can help (?).

**Comment by HarshNarayanJha** [comment]
RIght, forgot to add. I already saw that issue and downgraded `libappindicator-gtk3` to ` v12.10.0.r296-1`, with no success. Tray still disappears with the error if I restart waybar.

![image](https://github.com/user-attachments/assets/84a1fbc7-0574-4aba-8ccc-0313d4ef5fc4)

**Comment by careb0t** [comment]
Same issue here unfortunately. I seem to get the `Host: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: No such object path '/StatusNotifierWatcher'` error and thus the tray disappearing whenever my lock screen activates, my computer sleeps, or whenever I reload waybar because I updated a config file or what have you.

**Comment by Pierceisgone** [comment]
I've been having a similar issue here that's been going on for a month of two now (~Since I started using Hyprland).

I am using Hyprland on Arch. If I suspend/sleep my desktop, and then wake it again later, the tray no longer works. All the icons in it disappear, and programs can no longer run in the tray. E.g. if I play an audio file with VLC, instead of VLC starting in the tray, it'll open as a window. To fix this I have to reboot my PC (Although just restarting Hyprland may fix this but I haven't tested that yet). After the tray breaks, nothing can run trayed. Also, the tray does *not* break *every* time I suspend, but it does *most* of the time.

I am also getting the `error] Host: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: No such object path '/StatusNotifierWatcher'` error.

Hyprland version: hyprland 0.41.2-3
Waybar Version: waybar 0.10.4-1
Kernel Version: 6.10.0-arch1-2

My Waybar config: https://pastebin.com/srqbPLEq

I also ran Waybar in a terminal a for a while. I suspended my PC for ~5-8 minutes and then the tray broke when I woke it. These are the logs I got after waking the desktop: https://pastebin.com/0U6ZeECT

**Comment by ShikharArvind** [comment]
> * Similar issue on Sway + Waybar. Resuming after extended suspend makes the tray part of bar disappear. Killing and restarting it does not help.
> 
>     * Starting from terminal throws the error `Host: GDBus.Error:org.freedesktop.DBus.Error.UnknownObject: No such object path '/StatusNotifierWatcher'`
> 
>     * Restarting Sway itself fixes the issue, and the tray seems to be working fine.
> 
>     * Could be due to `libappindicator-gtk3` as described in issue [Tray vanishing at random instances #3040](https://github.com/Alexays/Waybar/issues/3040) .  Version is 12.10.0.r298-4 for me.  Maybe downgrading to 12.10.0.r296 can help (?).
> 
>     * Edit : Tried uninstalling  `libappindicator-gtk3` completely and still does not seem to resolve the issue.
> 
> 
> Waybar : v0.10.4 SwayWM on CachyOS (Arch) Kernel : 6.10.1-3-cachyos

I also have Plasma 6 installed along with Sway. I suspect this has something to do with 'org.kde.StatusNotifierWatcher' (ref : https://www.reddit.com/r/qtile/comments/17nx6qt/unable_to_get_status_notifier_working_on_qtile/k7xzvmc/ )

**Comment by ShikharArvind** [comment]
I've been trying to debug this issue and I think I have a better idea now. I've Plasma 6 installed along with Sway + Waybar.  Here is what I've found so far, maybe this will help others facing similar issue. 

- During the startup Waybar process owns `org.kde.StatusNotifierWatcher` and registers an `org.kde.StatusNotifierHost`. I can confirm this by running `busctl --user list|grep -i status`. The output is as follows:
```
org.kde.StatusNotifierHost-1358-1                                  1358 waybar          shikhar :1.8          session-3.scope   3       -
org.kde.StatusNotifierWatcher                                      1358 waybar          shikhar :1.8          session-3.scope   3       -
````

- After and extended suspend and resume, the tray from Waybar disappears. If I run `busctl -- user list|grep -i status` again, its seems like `kded6` now owns `org.kde.StatusNotifierWatcher`.
```
org.kde.StatusNotifierHost-1358-1                                  1358 waybar          shikhar :1.8          session-3.scope   3       -
org.kde.StatusNotifierWatcher                                     17162 kded6           shikhar :1.73         user@1000.service -       -
```
-  Killing `kded6` us `pkill` brings back the tray in Waybar , so this definitely seems to have something to do with this error. But I checked `journalctl` and `kded6` was already running before the suspend and resume, and the tray was working fine then. 
-  Since the unit for `kded6` was `user@1000.service`, I ran `systemctl status user@1000.service` and these were the relevant parts:
```
Jul 31 06:38:37  kactivitymanagerd[16997]: qt.qpa.wayland: Creating a fake screen in order for Qt not to crash
Jul 31 06:38:37  kded6[17162]: qt.qpa.wayland: Creating a fake screen in order for Qt not to crash
```
- Whenever I boot, `kactivitymanagerd` and `kded6` are not running in on my system. Only when I open `Dolphin` ( and presumably other KDE applications),`kactivitymanagerd` starts via D-bus service. (Probably something similar  triggers the starting of `kded6`).
- Even if they are running in the background, as long as my computer is not suspended, the tray seems to work fine. Waybar still owns `org.kde.StatusNotifierWatcher` .   Only when I suspend for long and resume, the tray disappears. Presumably due to whatever this does `kded6[17162]: qt.qpa.wayland: Creating a fake screen in order for Qt not to crash` and `kded6` is now the process owning `org.kde.StatusNotifierWatcher` .
- A very hacky fix for this : the D-bus services for starting `kactivitymanagerd` and `kded6` are located in  `/usr/share/dbus-1/services`. I renamed both from `org.kde.ActivityManager.service` and `org.kde.kded6.service` to `org.kde.ActivityManagerservice.backup` and `org.kde.kded6service.backup`. 
- `Dolphin` works fine with these modifications, but I have not extensively tested all the KDE apps. 
- Of course, this is not ideal and maybe be application  breaking if I switch to Plasma. For now it seems to work fine on Sway + using KDE applications. I will update this post incase something major breaks because of this.

**Comment by HarshNarayanJha** [comment]
Amazing debugging skills @ShikharArvind 👏🏼👏🏼👏🏼 ! I myself have been trying to debug this for some time, but couldn't figure out for myself how kde6 owns the tray after killing waybar and/or suspending, and how to give ownership back to waybar.

I also have KDE Plasma alongside hyprland, and this must be the issue.

For the solution, is there some mechanism to ask for the ownership of a dbus service? AFAIK, there is not, since all notification daemons ask to kill any running notification daemon. Maybe waybar can check and kill `kded6` at launch?

**Comment by ShikharArvind** [comment]
Thanks! I'm out of my depth when it comes to C++ or just general D-Bus spec, so not sure if there is an elegant/idiomatic way to check for this by waybar itself. This is something for more knowledge people to check. In the mean time, I did come up with a quick and dirty workaround for this : https://github.com/ShikharArvind/check_waybar_tray . It seems to work fine for me but I have not tested it all that much.

**Comment by HarshNarayanJha** [comment]
Seems good! I will check it out later

**Comment by aziz-204** [comment]
@ShikharArvind thank you so much bro <3 but i wanna ask how did you made this program bcz i'm not a good programmer and i want to learn stuff if you can help. ty

**Comment by mt190502** [comment]
> EDIT : Disabling `org.kde.ActivityManager.service` and `org.kde.kded6.service` definitely breaks Plasma if you ever want to switch from WM to it.

This situation happened to me too. But when I look, I cannot find these services in my system. One of the apps I use must be triggering kded6

OS: Fedora 41
Waybar version: v0.11.0
Indicator versions:
```
libappindicator-gtk3           12.10.1
libindicator-gtk3              12.10.1
libappindicator-sharp          12.10.1
libappindicator                12.10.1
libindicator                   12.10.1
```

**Comment by mt190502** [comment]
> > EDIT : Disabling `org.kde.ActivityManager.service` and `org.kde.kded6.service` definitely breaks Plasma if you ever want to switch from WM to it.
> 
> This situation happened to me too. But when I look, I cannot find these services in my system. One of the apps I use must be triggering kded6
> 
> OS: Fedora 41 Waybar version: v0.11.0 Indicator versions:
> 
> ```
> libappindicator-gtk3           12.10.1
> libindicator-gtk3              12.10.1
> libappindicator-sharp          12.10.1
> libappindicator                12.10.1
> libindicator                   12.10.1
> ```

I uninstalled the KDE applications I was using along with their dependencies. And reinstalled this apps (dolphin, kate, okular etc.) without weak kde plasma dependencies. Problem solved.

**Comment by tomektom** [comment]
From my experiments (mainly trying to use _Open as administrator_ in Dolphin provided by `kio-admin`, also Dolphin's builtin file search seems to have this issue) seems it's caused exclusively by kded6 dbus service – so I disabled it as described on [Arch Wiki](https://wiki.archlinux.org/title/D-Bus#Override_dbus_service) and that solved everything (for now). KDE also starts with this dbus disabled (but I didn't test much here)

Instead of disabling permanently and facing unknown problems in KDE, I disabled kded6 only for Hyprland with these two lines, kded6 never start by itself :smile: 
```
exec-once = printf "[D-BUS Service]\nName=org.kde.kded6\nExec=/bin/false" > $HOME/.local/share/dbus-1/services/org.kde.kded6.service
exec-shutdown = rm $HOME/.local/share/dbus-1/services/org.kde.kded6.service
```

**Comment by ShikharArvind** [comment]
@tomektom Ooh, I like this approach a lot better than what I did. **_I still have to test it and see if it works for me too._** 

For those wondering how to implement this for sway, it is quite similar to hyprland except there is no exec-shutdown so you need to append the `rm` command to your sway exit bind. Here  are the relevant lines in my sway config :

```
exec printf "[D-BUS Service]\nName=org.kde.kded6\nExec=/bin/false" > $HOME/.local/share/dbus-1/services/org.kde.kded6.service

bindsym $mod+Shift+e exec swaynag -t warning -m 'You pressed the exit shortcut. Do you really want to exit sway? This will end your Wayland session.' -B 'Yes, exit sway' 'rm -f $HOME/.local/share/dbus-1/services/org.kde.kded6.service && swaymsg exit'
```

I used `rm -f` because it will not throw any error if the file doesn't exist and move ahead with the `swaymsg exit`.

**Comment by Set2Minecraft** [comment]
Had this exact issue, killing kde6 solved it.

Waybar v0.12.0
Sway 1.10.1

**Comment by afrolino02** [comment]
> Had this exact issue, killing kde6 solved it. (edit: for now)
> 
> Waybar v0.12.0 Sway 1.10.1

You saved my live dude, thanks!

**Comment by klaasb** [comment]
I had the same issue with swaybar, so I investigated a bit more.
In theory, kded6 is configurable and the statusnotifierwatcher module could be disabled from loading via `~/.config/kded5rc` (yes, 5...) - which should be the least intrusive way to only disable the conflicting part:
```
[Module-statusnotifierwatcher]
autoload=false
```
However, kded [always registers for all DBus names of any module](https://invent.kde.org/frameworks/kded/-/blob/42b5d60bad5021d47b7ab0920c4c126584d5af81/src/kded.cpp#L606), even if it's prohibit to load, so as soon as our bars/systray apps disappear for some reason - kded gets it even if it doesn't actually have the module loaded for it.

So, one possible fix would be in kded6 to not grab these names (and get into the queue to grab them as well!) if the module shouldn't be loaded - then this could be actually configurable with existing methods.

Another possible fix would be - as far as I understand DBus and that's not a lot - for the bar/tray to register with `REPLACE_EXISTING` flags to DBus and make sure that whenever happens whatever happens in these cases (suspend?) it re-registers with `REPLACE_EXISTING` to kick kded6 off - or make sure it doesn't lose the name in the first place under the root issue.

**Comment by jmickelin** [comment]
I think I found the core issue. Let me try to summarize here.

TL;DR: Waybar _is_ already doing the right thing. The problem is due to an issue with the way Kded loads its modules when running outside of a KDE session. This has the effect of unregistering all applications that previously created tray icons. Forcing Kded to load the `statusnotifierwatcher` module, and then restarting those applications does make them show up again in Waybar. 

The best option is _still_ to just remove/disable Kded in non-KDE sessions. So nothing new there. But I hope this writeup irons out the remaining details of the mystery, for those interested.

# The `org.kde.StatusNotifierItem` / `org.freedesktop.StatusNotifierItem` DBus interfaces
(These two interfaces should be identical except for the name, though some programs still use the `org.kde.*` name for legacy reasons.)

The tray side of this spec consists of two halves, each with a different interface:
- [`StatusNotifierWatcher`](https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/StatusNotifierWatcher/): Controls a service that listens for applications that want to create tray icons, and keeps track of the state for those. It also acts as a registration service for any number of `StatusNotifierHost`s. Notably, it will unregister all application icons when the last `StatusNotifierHost` shuts down.
- [`StatusNotifierHost`](https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/StatusNotifierHost/): Interface that can be used by any number of "backend" programs. These are the programs that actually render the trays.

Lots of programs (Waybar included) can act as both a `Watcher` and a `Host`. There can be only a single `Watcher`, but that isn't normally a problem: Waybar works fine even if it ends up using a different `Watcher` service.

Kded is the same. Well, sort of...

# Kded internals

Here is what the bus looks like when Waybar is acting as the `Watcher`:
```bash
$ busctl --user tree org.kde.StatusNotifierWatcher
├─ /StatusNotifierWatcher
├─ /fr
│ └─ /fr/arouillard
│   └─ /fr/arouillard/waybar
└─ /org
  └─ /org/gtk
    └─ /org/gtk/Profiler      
```
The important part is the `/StatusNotifierWatcher` object, which implements the `org.kde.StatusNotifierWatcher` interface. You can confirm this by running

```bash
$ busctl --user introspect org.kde.StatusNotifierWatcher /StatusNotifierWatcher
```

Now compare it to the situation when Kded owns the bus (turn off Waybar and start `kded6` manually to force it to get to this state):

```bash
$ busctl --user tree org.kde.StatusNotifierWatcher
├─ /MainApplication
├─ /kbuildsycoca
├─ /kded
└─ /org
  └─ /org/kde
    └─ /org/kde/kded6
```

_Weird_! So it looks like Kded hogs the bus, as discussed above. But even worse: It doesn't actually implement the interface at all! There is no service listening when applications send requests to create tray icons.

Why is that?

It turns out that the answer lies in the `/kded` object. You can send a request like this:

```bash
$ busctl --user call org.kde.StatusNotifierWatcher /kded org.kde.kded6 loadedModules 
as 0
```

and observe that we get an empty array as return value. So it looks like the `statusnotifierwatcher` module isn't being loaded. You can force it to do so:

```bash
$ busctl --user call org.kde.StatusNotifierWatcher /kded org.kde.kded6 loadModule s statusnotifierwatcher
b true

$ busctl --user call org.kde.StatusNotifierWatcher /kded org.kde.kded6 loadedModules
as 1 "statusnotifierwatcher"
```

Now, if you try to restart Waybar, it still won't show any icons. But then (re)start any program that creates an icon. Those icons _do_ show up in the Waybar tray. So, while no icons that were created before Kded took over will come back, newly created ones work normally.

You can also manually register Waybar as a host with the Kded `Watcher`:

```bash
$ busctl --user status org.kde.StatusNotifierHost-260288-1
[...]
Comm=waybar
[...]

$ busctl --user call org.kde.StatusNotifierWatcher /StatusNotifierWatcher org.kde.StatusNotifierWatcher RegisterStatusNotifierHost s org.kde.StatusNotifierHost-260288-1
```

# Explanation
You can trace what happens during Kded's startup by running `busctl --user monitor  org.kde.StatusNotifierWatcher`, but here is the summary:

1. Kded registers itself as a `Host`, and broadcasts this on the entire bus.
2. Client applications receive this broadcast and send out their icon requests to `/StatusNotifierWatcher`.
3. _Now_ Kded notices that it doesn't actually have the `statusnotifierwatcher` module loaded, so the object the clients are sending their message to doesn't exist. It barfs out a lot of "No such object path '/StatusNotifierWatcher'" errors.
4. It doesn't look like the clients are ever made aware of the fact that the `Host` that was broadcast isn't actually running.
5. Loading the `statusnotifierwatcher` does not cause a new broadcast to be sent.

So the core issue is that Kded doesn't autoload the module in question, and doing so through DBus after it has already started will leave the clients unaware.

# Why is the module not being loaded?
You can add an `autoload` line in the `kded5rc` file as mentioned by @klaasb (or use an equivalent DBus call `setModuleAutoloading sb statusnotifierwatcher true`).

But it won't actually do anything outside of KDE sessions! 😩 

Why? Well, in addition to the `autoload` attribute, there is _another_ one called `X-KDE-Kded-phase`, [with the following semantics](https://invent.kde.org/plasma/plasma-workspace/-/blob/dfa2e04e5149becc690169cae9eb837514537512/ksmserver/README):
> The exact way autoloading works is controlled by X-KDE-Kded-phase=,
which may be 0, 1 or 2 (the default). Kded phase 0 means the module is
always loaded by kded, even outside of KDE session. It should used only
by kded modules which must be always running. Kded phase 1 modules are
loaded right after kded startup, but only during KDE startup, i.e. it is
for modules that are always needed by the KDE session. Phase 2 modules
will be loaded later.

For `statusnotifierwatcher`, it is apparently set to `1`. So any configuration you set to autoload it will be ignored, unless you are running KDE.

There is no obvious way to override the module metadata. In fact, even reading it is a pain: It's stored as a chunk of JSON data inside of the module's `.so` file! (The linked documentation talks about `.desktop` files, but those aren't being used anymore.) You can take a peek at the module metadata by starting Kded like so:

```bash
$ QT_DEBUG_PLUGINS=1 kded6 
```

# Conclusions
This is pretty much as far as I got. It would be great if you could override the `X-KDE-Kded-phase` value and set it to `0`, but I have no hope of doing so inside a binary library (and without triggering my NixOS to rebuild all of the KDE suite).

I did try overriding some environment variables to trick Kded into thinking it was running a KDE session, but with no luck.

Even worse, this logic was removed entirely just a few weeks ago: 
- https://invent.kde.org/frameworks/kded/-/merge_requests/77
- https://invent.kde.org/frameworks/kded/-/work_items/3

Looking at the code, it does look like it's fully controlled by an `autoload` value now. But the comments in the PR imply that it's now working as if `phase` was set to `1` in the previous versions. I guess I'd have to try it to find out.

Also mentioned in the linked PR: The best option is probably to just never use Kded outside of KDE. (But it's not as if we have a choice when other applications are spawning it. It's also the thing responsible for running `kbuildsycoca` when needed, so completely removing it would break file associations and icons in e.g. Dolphin... So you'd have to replace it with your own `ksycoca` script.)

----

One final takeaway is that we can definitively conclude that this isn't actually a bug in Waybar. It keeps working as it should, provided the `Watcher` properly tells the applications that they can register icons. To confirm, you can start Waybar alongside another bar (I tried Ironbar). Whichever order you start them in (so whichever bar claims ownership of the bus), both bars will display the same icons.

That is not to say Waybar couldn't do workarounds to fix broken behavior from other programs, of course! I think I will just try patching my systemd `.service` file to kill Kded whenever Waybar starts. Not elegant, but what can you do.

**Comment by jmickelin** [comment]
Here is my workaround that builds upon the one from @ShikharArvind and @tomektom, but that should work in Plasma. This means it's a set-and-forget method, that saves you the need of using scripts to add/remove files to disable the DBus services.

1. Create the following DBus service file in `$HOME/.local/share/dbus-1/services/org.kde.kded6.service`:
  
   ```desktop
   [D-BUS Service]
   Name=org.kde.kded6
   SystemdService=plasma-kded6.service
   ```
   Here, `plasma-kded6.service` refers to a systemd service that was already present on my system. If yours doesn't have it, you can create one (see below).

   Do the same for `org.kde.ActivityManager.service` (though I found that on my system, it already has the `SystemdService` line as needed):

   ```desktop
   [D-BUS Service]
   Name=org.kde.ActivityManager
   SystemdService=plasma-kactivitymanagerd.service
   ```

2. Create a drop-in file in `$HOME/.config/systemd/user/plasma-kded6.service.d/overrides.conf` (create the directory if it doesn't exist):

   ```service
   [Service]
   ExecCondition=/usr/lib/systemd/user-generators/systemd-xdg-autostart-condition "KDE" ""
   ```

   Do the same for `$HOME/.config/systemd/user/plasma-kactivitymanagerd.service.d/overrides.conf`.

   You might have to adjust the path where `systemd-xdg-autostart-condition` is located, depending on your distro.

The trick is to use `SystemdService` instead of `Exec=/bin/false`. So rather than completely disabling the DBus services, we delegate the program activation to systemd. The benefit in doing so is that systemd provides _a lot_ more machinery for controlling service dependencies. The check I added will succeed in a KDE session, but fail otherwise.

But `ExecCondition=` and `systed-xdg-autostart-condition` is just a suggestion (one that has worked well for me in the past) and you can do other things depending on your needs. For example you could use the options in [systemd.unit](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html) to make it stop as soon as the Waybar service starts or even mask the unit entire (disables it completely like before, but in a more discoverable way than creating/removing files), to name a few.

---
For reference, here is the contents of my existing systemd services. I think they are the defaults from upstream, but just in case:

**plasma-kded6.service**:
<details>

```service
[Unit]
Description=KDE Daemon 6
PartOf=graphical-session.target
After=plasma-kcminit.service

[Service]
ExecStart=/usr/bin/kded6
BusName=org.kde.kded6
Slice=session.slice
TimeoutStopSec=5sec
Restart=on-failure
```

</details>

**plasma-kactivitymanager.service**:
<details>

```service
[Unit]
Description=KActivityManager Activity manager Service
PartOf=graphical-session.target
Before=graphical-session.target

[Service]
ExecStart=/usr/libexec/kactivitymanagerd
BusName=org.kde.ActivityManager
Slice=background.slice
```

</details>

Again, not sure what the exact `ExecStart=` paths would be on other distros.

---

For my fellow NixOS users out there, here is my config to do the above in a Nix-y way:

<details>

```nix
services.dbus.packages = [
  (pkgs.writeTextFile {
    name = "kded6-dbus-activation-file";
    text = ''
      [D-BUS Service]
      Name=org.kde.kded6
      SystemdService=plasma-kded6.service
    '';
    destination = "/share/dbus-1/services/org.kde.kded6.service";
  })
];

systemd.user.services.plasma-kded6 = {
  overrideStrategy = "asDropin";
  serviceConfig = {
    ExecCondition = ''${pkgs.systemd}/lib/systemd/systemd-xdg-autostart-condition "KDE" ""'';
  };
};
systemd.user.services.plasma-kactivitymanagerd = {
  overrideStrategy = "asDropin";
  serviceConfig = {
    ExecCondition = ''${pkgs.systemd}/lib/systemd/systemd-xdg-autostart-condition "KDE" ""'';
  };
};
```

</details>

**Comment by rhoot** [comment]
I ran into this project last week which seems to work as a workaround: https://github.com/wsnrq/sni-guard
